Mathlib/Data/Matroid/Rank/Cardinal.lean

```diff
@@ -63,40 +136,120 @@ theorem Base.cardinalMk_diff_comm_of_finitary (hB : M.Base B) (hB' : M.Base B')
-    #(I \ J : Set α) = #(J \ I : Set α) :=
-  hIX.basis'.cardinalMk_diff_comm_of_finitary hJX.basis'
-
-theorem Basis'.cardinalMk_eq_of_finitary (hIX : M.Basis' I X) (hJX : M.Basis' J X) : #I = #J := by
-  rw [← base_restrict_iff'] at hIX hJX
-  rw [hIX.cardinalMk_eq_of_finitary hJX]
-
-theorem Basis.cardinalMk_eq_of_finitary (hIX : M.Basis I X) (hJX : M.Basis J X) : #I = #J :=
-  hIX.basis'.cardinalMk_eq_of_finitary hJX.basis'
-
-theorem Indep.cardinalMk_le_base_of_finitary (hI : M.Indep I) (hB : M.Base B) : #I ≤ #B :=
-  have ⟨_B', hB', hIB'⟩ := hI.exists_base_superset
-  hB'.cardinalMk_eq_of_finitary hB ▸ mk_le_mk_of_subset hIB'
-
-theorem Indep.cardinalMk_le_basis'_of_finitary (hI : M.Indep I) (hJ : M.Basis' J X) (hIX : I ⊆ X) :
-    #I ≤ #J :=
-  have ⟨_J', hJ', hIJ'⟩ := hI.subset_basis'_of_subset hIX
-  hJ'.cardinalMk_eq_of_finitary hJ ▸ mk_le_mk_of_subset hIJ'
-
-theorem Indep.cardinalMk_le_basis_of_finitary (hI : M.Indep I) (hJ : M.Basis J X) (hIX : I ⊆ X) :
-    #I ≤ #J :=
-  hI.cardinalMk_le_basis'_of_finitary hJ.basis' hIX
+/-- Restrictions of matroids with cardinal rank functions have cardinal rank functions- -/
+instance cardinalRank_restrict [CardinalRank M] : CardinalRank (M ↾ X) := by
+  refine ⟨fun I J Y hI hJ ↦ ?_⟩
+  rw [basis_restrict_iff'] at hI hJ
+  exact hI.1.cardinalMk_diff_comm hJ.1
+
+instance cardinalRank_map {α β : Type u} {f : α → β} (M : Matroid α) [CardinalRank M]
+    (hf : InjOn f M.E) : CardinalRank (M.map f hf) := by
+  refine ⟨fun I J X hI hJ ↦ ?_⟩
+  obtain ⟨I, X, hIX, rfl, rfl⟩ := map_basis_iff'.1 hI
+  obtain ⟨J, X', hJX, rfl, h'⟩ := map_basis_iff'.1 hJ
+  obtain rfl : X = X' := by
+    rwa [InjOn.image_eq_image_iff hf hIX.subset_ground hJX.subset_ground] at h'
+  have hcard := hIX.cardinalMk_diff_comm hJX
+  rwa [← mk_image_eq_of_injOn _ _ (hf.mono (diff_subset.trans hIX.indep.subset_ground)),
+    ← mk_image_eq_of_injOn _ _ (hf.mono (diff_subset.trans hJX.indep.subset_ground)),
+    (hf.mono hIX.indep.subset_ground).image_diff,
+    (hf.mono hJX.indep.subset_ground).image_diff, inter_comm,
+    hf.image_inter hJX.indep.subset_ground hIX.indep.subset_ground,
+    diff_inter_self_eq_diff, diff_self_inter] at hcard
+
+end Instances
+
+section Rank
+
+/-- The rank (size of a base) of a matroid `M` as a `Cardinal`. -/
+def cardRank (M : Matroid α) := #(M.exists_base.choose)
+
+/-- The rank (size of a basis) of a set `X` in a matroid `M`, as a `Cardinal`. -/
+def cardRk (M : Matroid α) (X : Set α) := (M ↾ X).cardRank
+
+variable [CardinalRank M]
+
+theorem Base.cardRank_eq_cardinalMk (hB : M.Base B) : M.cardRank = #B :=
+  M.exists_base.choose_spec.cardinalMk_eq hB
+
+theorem Basis'.cardRk_eq_cardinalMk (hIX : M.Basis' I X) : M.cardRk X = #I :=
+  (M ↾ X).exists_base.choose_spec.cardinalMk_eq (base_restrict_iff'.2 hIX)
+
+theorem Basis.cardRk_eq_cardinalMk (hIX : M.Basis I X) : M.cardRk X = #I :=
+  hIX.basis'.cardRk_eq_cardinalMk
```

Comment:
I think it makes more sense to swap the two sides of the identities, in conformity with [Basis.mk_eq_rank''](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Dimension/StrongRankCondition.html#Basis.mk_eq_rank'') etc. Then these could potentially be simp lemmas. (The current direction can't be because there can be multiple RHS for the same LHS.)

Reply:
I think I disagree (conceptually, 'rank' is more complicated than 'size'). And it feels like it's the wrong direction for simplification in most cases, as well.

Reply:
`rank` is defined to be the size of a random base, right? So size is more fundamental than `rank`. And size -> rank is conceptually the "normalizing" direction, as it turns potentially different sizes into the same size.

Reply:
I see your point, and that had never occured to me. I will change them.